File: handlers/scorevspt.go
package handlers

import (
	"encoding/json"
	"errors"
	"fmt"
	commonmodels "github.com/dialangproject/common/models"
	"github.com/dialangproject/web/data"
	"github.com/dialangproject/web/datacapture"
	"github.com/dialangproject/web/models"
	"github.com/dialangproject/web/session"
	"log"
	"math"
	"net/http"
	"strings"
)

func ScoreVSPT(w http.ResponseWriter, r *http.Request) {

	var vsptResponses map[string]string
	decoder := json.NewDecoder(r.Body)
	err := decoder.Decode(&vsptResponses)
	if err != nil {
		log.Fatal(err)
	}

	responses := map[string]bool{}
	for k, v := range vsptResponses {
		if strings.HasPrefix(k, "word:") {
			id := strings.Split(k, ":")[1]
			responses[id] = v == "valid"
		}
	}

	dialangSession := session.SessionManager.Get(r.Context(), "session").(models.DialangSession)

	zScore, mearaScore, level, err := getBand(dialangSession.TES.TL, responses)
	if err != nil {
		log.Fatal(err)
	}

	dialangSession.VsptZScore = zScore
	dialangSession.VsptMearaScore = mearaScore
	dialangSession.VsptLevel = level
	dialangSession.VsptSubmitted = true

	session.SessionManager.Put(r.Context(), "session", dialangSession)

	datacapture.LogVSPTResponses(&dialangSession, responses)
	datacapture.LogVSPTScores(&dialangSession)

	json.NewEncoder(w).Encode(map[string]any{"vsptZScore": zScore, "vsptMearaScore": mearaScore, "vsptLevel": level})
}

func getBand(tl string, responses map[string]bool) (float64, int, string, error) {

	zScore, mearaScore := getScore(tl, responses)
	bands, ok := data.VSPTBands[tl]
	if !ok {
		return 0, 0, "", errors.New(fmt.Sprintf("No bands for test language '%s'", tl))
	}

	for _, b := range bands {
		log.Println(b)
		if mearaScore >= float64(b.Low) && mearaScore <= float64(b.High) {
			return zScore, int(math.Round(mearaScore)), b.Level, nil
		}
	}

	return 0, 0, "", errors.New(fmt.Sprintf("No level for test language '%s' and meara score: %f.", tl, mearaScore))
}

func getScore(tl string, responses map[string]bool) (float64, float64) {

	Z := getZScore(tl, responses)

	log.Printf("Z Score: %f\n", Z)

	if Z <= 0 {
		return Z, 0
	} else {
		return Z, Z * 1000
	}
}

func getWordType(word *commonmodels.VSPTWord) int {

	var wordType int
	if word.Valid == 1 {
		wordType = 1
	}
	return wordType
}

func getZScore(tl string, responses map[string]bool) float64 {

	yesResponses := []int{0, 0}
	noResponses := []int{0, 0}

	words, ok := data.VSPTWords[tl]
	if !ok {
		log.Fatal("No vspt words for language", tl)
	}

	for _, word := range words {
		wordType := getWordType(&word)

		if responses[word.WordId] {
			yesResponses[wordType] += 1
		} else {
			noResponses[wordType] += 1
		}
	}

	realWordsAnswered := yesResponses[1] + noResponses[1]

	fakeWordsAnswered := yesResponses[0] + noResponses[0]

	// Hits. The number of yes responses to real words.
	hits := yesResponses[1]

	// False alarms. The number of yes responses to fake words.
	falseAlarms := yesResponses[0]

	if hits == 0 {
		// No hits whatsoever results in a zero score
		return 0
	} else {
		return getVersion10ZScore(hits, realWordsAnswered, falseAlarms, fakeWordsAnswered)
	}
}

func getVersion10ZScore(hits int, realWordsAnswered int, falseAlarms int, fakeWordsAnswered int) float64 {

	h := float64(hits) / float64(realWordsAnswered)

	// The false alarm rate. False alarms divided by the total number of fake words answered.
	f := float64(falseAlarms) / float64(fakeWordsAnswered)

	if h == 1 && f == 1 {
		// This means the test taker has just clicked green for all the words
		return -1
	} else {
		rhs := ((4 * h * (1 - f)) - (2 * (h - f) * (1 + h - f))) / ((4 * h * (1 - f)) - ((h - f) * (1 + h - f)))
		return 1 - rhs
	}
}

File: src/routes/scorevspt.ts
import { getSessionId } from "../utils/utils.ts";
import type { Storage } from "../storage/storage.ts";
import type { DialangSession, VSPWord } from "../types.ts";
import type { Context } from "@hono";

export async function scoreVspt(
  c: Context,
  storage: Storage,
): Promise<Response> {

  const body = await c.req.json();

  const responses = Object.fromEntries(
    Object.entries(body)
    .filter(([k, v]) => k.startsWith("word:"))
    .map(([k, v]) => [ k.split(":")[1], v === "valid"])
  );

  const sessionId = getSessionId(c);
  if (!sessionId) {
    console.error("Failed to get session id");
		c.status(500);
		return c.html("");
  }

  const session: DialangSession | null = await storage.getSession(sessionId);
  if (!session) {
		console.error(`No session for id ${sessionId}. Returning 500 ...`)
		c.status(500);
		return c.html("");
  }

  if (!session.tl) {
		console.error("Test language not set. Returning 500 ...")
		c.status(500);
		return c.html("");
  }

  const [ zScore, mearaScore, level, error ] = await getBand(storage, session.tl, responses);
  if (error) {
    c.status(500);
    return c.text(error as string);
  }

  session.vsptZScore = zScore as number;
  session.vsptMearaScore = mearaScore as number;
  session.vsptLevel = level as string
  session.vsptSubmitted = true

  storage.saveSession(sessionId, session);

  storage.logVsptScores(session);

  return c.json({ zScore, mearaScore, level });
}

async function getBand(storage: Storage, tl: string, responses: Record<string, boolean>): Promise<Array<number | string | null>> {

  const [zScore, mearaScore] = await getScore(storage, tl, responses);

  const bands = await storage.getVSPBands(tl);
  if (!bands) {
    return [ 0, 0, "", `No bands for test language '${tl}` ];
  }

  const match = bands.find(b => mearaScore >= b.low && mearaScore <= b.high);
  if (match) {
  return [ zScore, Math.round(mearaScore), match.level, null ];
  }

  return [ 0, 0, "", `No level for test language '${tl}' and meara score: ${mearaScore}.` ];
}

async function getScore(storage: Storage, tl: string, responses: Record<string, boolean>): Promise<Array<number>> {

  const Z = await getZScore(storage, tl, responses);

  if (Z <= 0) {
    return [ Z, 0 ];
  }

  return [Z, Z * 1000 ];
}

async function getZScore(storage: Storage, tl: string, responses: Record<string, boolean>): Promise<number> {

  const words: Array<VSPWord> | null = await storage.getVSPWords(tl);
  if (!words) {
    console.error(`No vspt words for test language ${tl}`);
    return -1;
  }

  const yesResponses = [0, 0];
  const noResponses = [0, 0];

  words.forEach(word => {

    const wordType = word.valid == 1 ? 1 : 0;

    if (responses[word.word_id]) {
      yesResponses[wordType] += 1;
    } else {
      noResponses[wordType] += 1;
    }
  });

  const realWordsAnswered = yesResponses[1] + noResponses[1];

  const fakeWordsAnswered = yesResponses[0] + noResponses[0];

  // Hits. The number of yes responses to real words.
  const hits = yesResponses[1];

  // False alarms. The number of yes responses to fake words.
  const falseAlarms = yesResponses[0];

  if (hits == 0) {
    // No hits whatsoever results in a zero score
    return 0;
  } else {
    return getVersion10ZScore(hits, realWordsAnswered, falseAlarms, fakeWordsAnswered);
  }
}

function getVersion10ZScore(hits: number, realWordsAnswered: number, falseAlarms: number, fakeWordsAnswered: number): number {

  const h = hits / realWordsAnswered;

  // The false alarm rate. False alarms divided by the total number of fake words answered.
  const f = falseAlarms / fakeWordsAnswered;

  if (h == 1 && f == 1) {
    // This means the test taker has just clicked green for all the words
    return -1;
  } else {
    const rhs = ((4 * h * (1 - f)) - (2 * (h - f) * (1 + h - f))) / ((4 * h * (1 - f)) - ((h - f) * (1 + h - f)));
    return 1 - rhs;
  }
}
